kicad-unstable-small: init to build kicad-unstable's base on hydra

evils opened the pull request:
###### Motivation for this change
`kicad-unstable` does not get built by hydra

###### Things done

* add `kicad-unstable-small` so the components of `kicad-unstable` that hydra can build, get build
* make the CLI utilities available by linking to them them in the wrapper
* move `libraries` and `base` out of the `let` block so they can be addressed directly
* bump kicad-unstable to latest versions

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

veprbl commented on pkgs/applications/science/electronics/kicad/libraries.nix:
```diff
@@ -35,7 +35,7 @@ in
   footprints = mkLib "footprints" { };
   packages3d = mkLib "packages3d" {
-    hydraPlatforms = []; # this is a ~1 GiB download, occupies ~5 GiB in store
+    meta.hydraPlatforms = []; # this is a ~1 GiB download, occupies ~5 GiB in store
```
Since we have
https://github.com/NixOS/nixpkgs/blob/f241aa6910450c3a54c8cf5bf0c501f72e09dcef/pkgs/applications/science/electronics/kicad/default.nix#L12
this line should probably be removed.

Evils-Devils replied:
the goal of that line is to not have the >5GB of 3d models on hydra,
it exceeds the output limit and it's just a download - unpack - move operation

i tagged you on this because you wrote the original line and i was hoping you'd know if this fix is correct

veprbl replied:
The goal is also to have the default `kicad` being built on Hydra without a failure. If we think that setting `meta.hydraPlatforms` will prevent Hydra from downloading packages3d, then it will also prevent `kicad` from being built there, which is not the best. This should is formally okay to merge for ZHF, of course.

Evils-Devils replied:
unfortunately i don't know how to achieve the desired effect.  
at least `kicad-base` should be built by hydra, but isn't because it's not user usable so i don't want it in all-packages.nix
everything else is just downloads and default.nix (`kicad` and `kicad-unstable`) just wraps base with libraries

if there was a way to make the sub-packages build on their own, i'd mark all but base package (and the 18n one (a direct dependency)) as hydraPlatforms = [];

as there is no build-time dependency on the content of `kicad-packages3d`, maybe there's another way of resolving this?

veprbl replied:
You could try to use `recurseIntoAttrs` for this.

Evils-Devils replied:
i'm not sure what that is a response to, i assume making the sub-packages build.  
i would kinda expect `nix repl` to then have `kicad.libraries.packages3d`, which it doesn't

veprbl commented:
@GrahamcOfBorg build kicad-unstable

veprbl reviewed (commented):
This is three different changes. Please split them into separate commits.

veprbl commented:
mkLib needs to be updated to use `recursiveUpdate`[1] so that `meta.license` doesn't get overwritten.

[1] https://github.com/NixOS/nixpkgs/blob/fbafe7ca64e9ce5dae908c6520035a0355043648/lib/attrsets.nix#L405-L427

Evils-Devils commented:
@veprbl i've changed it to use `recursiveUpdate`, it didn't affect the build output, i assume the metadata is stored separately, the libraries aren't directly addressable so i couldn't use `nix repl` to check if this worked...

veprbl commented on pkgs/applications/science/electronics/kicad/versions.nix:
```diff
@@ -27,25 +27,25 @@
   "kicad-unstable" = {
     kicadVersion = {
-      version =			"2020-02-10";
+      version =			"2020-03-16";
```
According to the guidelines, the commit message should be "kicad-unstable: 2020-02-10 -> 2020-03-16"

Evils-Devils replied:
right!

veprbl commented on pkgs/applications/science/electronics/kicad/update.sh:
```diff
@@ -175,6 +175,13 @@ done
 printf "}\n"
 } > "${tmp}"
 
+grep "\"\"" "${tmp}"
```
I guess, it would be more robust if the script was written assuming `set -e`. But this is a big script now...

Evils-Devils replied:
i'm fairly unfamiliar with `set -e`, looks like that wouldn't convey that the error is most likely caused by running out of space

veprbl replied:
Of course, `set -e` would not know that there is disk involved, so it won't provide informative error messages by itself. The point of using it is that shell scripts fail for other reasons and they better be failing immediately whenever something unexpected occurs.

Evils-Devils replied:
i've added `set -e`, waiting for a rebuild on latest revs and got to go afk for a while, maybe don't merge this until then

veprbl commented on pkgs/applications/science/electronics/kicad/update.sh:
```diff
@@ -175,6 +177,12 @@ done
 } > "${tmp}"
 
+empty_value="$(grep "\"\"" "${tmp}" || true)"
+if [[ -n $empty_value ]]; then
```
```suggestion
if grep "\"\"" "${tmp}"; then
```

veprbl commented on pkgs/applications/science/electronics/kicad/update.sh:
```diff
@@ -54,7 +56,7 @@ now=$(date --iso-8601)
 # just in case this runs in parallel
-rand="$(head -c 3 /dev/urandom | base64)"
+rand="$(cat /dev/urandom | tr -cd 'a-fA-F0-9' | head -c 5)"
 tmp="${here}/,versions.nix.${rand}"
```
```suggestion
tmp="${here}/,versions.nix.$RANDOM"
```

evils committed:
kicad-unstable-small: init to make kicad-unstable

the -small packages depend on all hydra buildable dependencies
the non-small ones depend on packages3d which exceeds hydra's limit

set platforms to all (kicad is cross-platform)
clarify package differences in the description

set maintainers on just the top level derivation

switch -unstable to not save debug symbols

indicate patch in version string

note broken dependencies

evils committed:
kicad: expose base and libraries attributes

Mic92 commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -4,11 +4,12 @@
 , ngspiceSupport ? true, libngspice
 , scriptingSupport ? true, swig, python3, python3Packages
-, debug ? false, valgrind
+, debug ? (!stable), valgrind
```
We usually don't build debug symbols on hydra. One can do though in an override.

Evils-Devils replied:
do you mean hydra somehow interferes with the compilation / forces stripping?

Mic92 replied:
No. It is more a convention that we don't ship debug symbols in nixpkgs by default because of the size overhead.

Evils-Devils replied:
considering the debug symbols are 750M and our -unstable won't be nightly, i'm willing to default debug to false

does it still make sense to `separateDebugInfo` then?
i haven't managed to use the separate debug symbols from an nix-env install

Mic92 replied:
I don't think it provides any benefits in that case.
Gdb would usually pick up debug symbols from alternative (configured) path.
I think I used it at some point and the documentation was helpful for that: https://sourceware.org/gdb/current/onlinedocs/gdb/Separate-Debug-Files.html

Mic92 commented:
cc @nek0 for testing.

nek0 commented:
Will try as soon as I can.

veprbl commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -35,17 +35,16 @@ let
+in
+stdenv.mkDerivation rec {
+
+  passthru.libraries = callPackages ./libraries.nix versionConfig.libVersion;
```
I don't think this is going to work. My understanding is that "nix-env" does not recurse into a derivation as if it was an attrset, neither will our release jobset (see https://github.com/NixOS/nixpkgs/pull/82760#issue-389547299 on how I was checking it there).

veprbl replied:
I mention this because later you add `hydraPlatforms` to `mkLib`. Otherwise, I think exposing libraries, by itself, is a step in the right direction towards making things more overridable.

Evils-Devils replied:
i added `hydraPlatforms` to `mkLib` to guard against attempting to build `packages3d` on hydra in case it's ever figured out to make the components build on their own.
i'm not sure how `nix-env` relates to this.
your (approx.) line from 82760: `echo ":p { inherit kicad-small kicad-unstable-small; }" | nix repl ./pkgs/top-level/release.nix`
reveals wxwidgets-3.1.2 and 3.0.4 are not supported on `x86_64-darwin`, but otherwise seems to evaluate.

Evils-Devils replied:
i've got it evaluating without error for darwin now
i don't have access to a darwin machine, but i'll try the same configuration on nixos tomorrow
`wxPython_4_0` doesn't account for `wxGTK` being `wxmac` on darwin
`libngspice` is currently set to `platforms.linux` though it should support darwin

veprbl replied:
I just wanted to make sure that you understand that release.nix will not discover `kicad.libraries.footprints` on its own and that none of our tools will check its `hydraPlatforms` for it either.

Regarding the darwin support, you, of course, don't need to implement it if you can't test or use it.

Evils-Devils commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -123,23 +122,33 @@ stdenv.mkDerivation rec {
+    # adjusting this package to downgrade to python2Packages.wxPython (wxPython 3),
+    # seems like more trouble than fixing wxPython_4_0 would be
+    # additionally, libngspice is marked as linux only, though it should support darwin
+    broken = stdenv.isDarwin;
```
I'm not sure this should actually be marked as broken, the issue currently seems to be dependencies and not this package...

Evils-Devils replied:
tag @tbenst since `wxPython_4_0` should support darwin, but doesn't use wxmac (and no maintainer is listed in the package)
and @bgamari since `libngspice` seems like [it should](https://github.com/imr/ngspice/blob/master/build-for-mac-os.sh) support darwin as well

evils replied:
asked on IRC, not marking this as broken until an actual darwin build starts and fails

evils committed:
kicad: link utils; clean up wrapper installPhase

evils committed:
kicad-unstable: 2020-04-25 -> 2020-05-06

Mic92 commented:
Result of `nixpkgs-review pr 82634` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
<br>- kicad-small
<br>- kicad-unstable-small
</details>

veprbl commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -56,9 +56,11 @@ stdenv.mkDerivation rec {
     substituteInPlace CMakeModules/KiCadVersion.cmake \
-      --replace "unknown" ${builtins.substring 0 10 src.rev}
+      --replace "unknown" "${builtins.substring 0 10 src.rev}-1" \
+      --replace "${version}" "${version}-1"
```
We don't use patchlevels in nixpkgs because the build process of a package can be amended in many different ways, and all those changes can be tracked by looking at the store paths hashes and inside drv files. Patchlevels can't keep up with they, they will quickly become outdated.

evils replied:
my goal with this was to make it clear in a kicad bug report (which contains that string) which commit the package was based on, while still making it clear a patch was added

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/kicad-suddenly-getting-rebuilt-from-source/16315/8
